src/observable.rs

```diff
@@ -0,0 +1,75 @@
+use crate::Sealed;
+use core::marker::PhantomData;
+
+pub trait IntoObservationToken: Sized + crate::Sealed {
+    type Peripheral;
+    fn into_ot(self) -> ObservationToken<Self::Peripheral>;
+}
+
+/// A struct to hold peripherals which are to be observed.
+///
+/// This prevents the observed peripheral from being consumed. Thus
+/// preventing things like a an observed gpio pin changing mode or an opamp from
+/// being disabled. This makes sure the underlaying peripheral will not
+/// change mode into something that is not compatible with what ever may be observing it.
+pub struct Observed<P, const OBSERVER_COUNT: usize> {
+    peripheral: P,
+}
+
+impl<P, const OBSERVER_COUNT: usize> Observed<P, OBSERVER_COUNT> {
+    /// Release the observation of this peripheral
+    ///
+    /// This returns the underlaying perpheral type. Since it is no longer
+    /// observed, you are once again free to do what you want with it.
+    pub fn release(self, _data: [ObservationToken<P>; OBSERVER_COUNT]) -> P {
+        self.peripheral
+    }
+}
+
+/// A struct to represent a registered observation of a peripheral of type `P`
+///
+/// The existence of this type guarantees that the observed peripheral will not
+/// change mode into something that is not compatibe with what ever is observing it
+pub struct ObservationToken<P> {
+    _p: PhantomData<P>,
+}
+
+/// A trait providing an interface to make peripherals observed
+///
+/// See [`Observed`] and [`ObservationToken`]
+pub trait Observable: Sized {
+    fn observe<const N: usize>(self) -> (Observed<Self, N>, [ObservationToken<Self>; N]) {
+        (
+            Observed { peripheral: self },
+            core::array::from_fn(|_| ObservationToken { _p: PhantomData }),
```

Comment:
I did some testing and determined this does not introduce any overhead. I recall that being a concern.

Reply:
Nice, thanks!